VM/vm_params2vm.py

```diff
@@ -0,0 +1,268 @@
+
+    Parameters
+    ----------
+    nzvm_cfg_path : Path to nzvm.cfg (initially placed in ptemp, and relocated after NZVM run is successful)
+    vm_params_dict : Dictionary containing relevant data
+    vm_dir : Directory where NZVM binary outputs are placed (cannot exist)
+    logger :
+    """
+    if nzvm_cfg_path is not None:
+        with open(nzvm_cfg_path, "w") as vmd:
+            vmd.write(
+                "\n".join(
+                    [
+                        "CALL_TYPE=GENERATE_VELOCITY_MOD",
+                        f"MODEL_VERSION={vm_params_dict['model_version']}",
+                        f"OUTPUT_DIR={vm_dir}",
+                        f"ORIGIN_LAT={vm_params_dict['MODEL_LAT']}",
+                        f"ORIGIN_LON={vm_params_dict['MODEL_LON']}",
+                        f"ORIGIN_ROT={vm_params_dict['MODEL_ROT']}",
+                        f"EXTENT_X={vm_params_dict['extent_x']}",
+                        f"EXTENT_Y={vm_params_dict['extent_y']}",
+                        f"EXTENT_ZMAX={vm_params_dict['extent_zmax']}",
+                        f"EXTENT_ZMIN={vm_params_dict['extent_zmin']}",
+                        f"EXTENT_Z_SPACING={vm_params_dict['hh']}",
+                        f"EXTENT_LATLON_SPACING={vm_params_dict['hh']}",
+                        f"MIN_VS={vm_params_dict['min_vs']}",
+                        f"TOPO_TYPE={vm_params_dict['topo_type']}\n",
+                    ]
+                )
+            )
+        logger.debug(f"Saved nzvm config file to {nzvm_cfg_path}")
+    else:
+        logger.debug("Path to nzvm.conf should be specified")
+
+
+def main(
+    name: str,
+    vm_params_path: Path,
+    outdir: Path,
+    vm_threads=1,
+    logger: Logger = qclogging.get_basic_logger(),
+):
+    """
+    Orchestrates conversion from vm_params.yaml to VM output
+
+    Parameters
+    ----------
+    name : name of the fault
+    vm_params_path : Path to the vm_params.yaml
+    outdir : Directory where output files are eventually saved (eg. ..../Data/VMs/Hossack)
+    vm_threads : Number of threads to run NZVM binary with if OpenMP-enabled
+    logger :
+
+    """
+
+    # temp directory for current process
+    with TemporaryDirectory(prefix=f"_tmp_{name}_", dir=outdir) as ptemp:
+        ptemp = Path(ptemp)
+        qclogging.add_general_file_handler(
+            logger, outdir / f"vm_params2vm_{name}_log.txt"
+        )
+
+        vm_working_dir = ptemp / "output"
+        nzvm_cfg_path = ptemp / "nzvm.cfg"
+        temp_vm_params_path = ptemp / "vm_params.yaml"
+
+        with open(vm_params_path, "r") as f:
+            vm_params_dict = yaml.load(f, Loader=yaml.SafeLoader)
+
+        if vm_params_dict is not None:
+            # saves nzvm.cfg
+            save_nzvm_cfg(nzvm_cfg_path, vm_params_dict, vm_working_dir, logger=logger)
+            # makes a copy of vm_params.yaml
+            copyfile(vm_params_path, temp_vm_params_path)
```

Comment:
It may look like it. If user specifies vm_params.yaml somewhere else, this is still doing something useful.  As we want to have a copy of vm_params.yaml in the outdir,

Reply:
If we want a copy in the output directory we should copy it directly there, not move it to temp then back out
I don't think it will come up often that vm_params won't be in the VM output directory anyway

Reply:
I also think that copying it into the output directory may be a bit of an assumption. Ideally, we want these scripts to be as dumb / simple as possible so I think we should just leave it where it is. In theory the workflow should be able to handle itself with a VM params in Timbuktu.

Reply:
vm_params.yaml : We leave it where it is. We don't move it.  However, we just make a copy of it in the same location as the created vm. By doing so we can know what the vm binary data is about just by opening the vm_params.yaml sitting next to it. Why copying to tmp directory, then move it to outdir, instead of copying it from the original place to outdir?
By doing so, I don't have to separately execute "copyfile" when I move all the files.
`   
 for f in files_to_move:
        move(f, outdir / f.name)  # may overwrite
    logger.debug("Removing Log and Velocity_Model directories")
`

Reply:
But you're already separately executing "copyfile" here...

Reply:
Yeah, it's personal preference - I like to have all things at one place. BTW, as long as everyone agrees that there is a need to keep a copy of vm_params.yaml next to vm output files, I'm happy to change this.

Reply:
While I was attempting to accommodate @jpa198, I remembered why I chose this way. Doing copyfile later means if vm_params.yaml already exists, I have to catch `except shutil.SameFileError`. Copy to the temp directory and move can silently overwrite it.

Reply:
```
if src != dest
    if dest.is_file():
        dest.unlink()
    copyfile(src, dest)
```

Reply:
Done. Merry Christmas. :)

Reply:
I think that if a vm_params.yaml already exists in the output directory (that isn't the one that was specified) that should be an abortable error and tell the user to sort their folder / command out. I don't think it's right to arbitrarily delete either one or attempt to handle it - it's a situation that shouldn't occur so telling the user that an error has occurred is probably the best.

Reply:
:christmas_tree:  :gift:  :santa:

Reply:
> I think that if a vm_params.yaml already exists in the output directory (that isn't the one that was specified) that should be an abortable error and tell the user to sort their folder / command out. I don't think it's right to arbitrarily delete either one or attempt to handle it - it's a situation that shouldn't occur so telling the user that an error has occurred is probably the best.

That's a fair point

Reply:
Done it. Aborting if there is an existing vm_params.yaml in the outdir and it is not the same as the specified vm_params.yaml.